File: samples/java-valueentity-store/src/main/java/store/customer/domain/CustomerValueEntity.java
package store.customer.domain;

import com.google.protobuf.Empty;
import kalix.javasdk.valueentity.ValueEntityContext;
import store.customer.api.CustomerApi;

public class CustomerValueEntity extends AbstractCustomerValueEntity {

  public CustomerValueEntity(ValueEntityContext context) {}

  @Override
  public CustomerDomain.CustomerState emptyState() {
    return CustomerDomain.CustomerState.getDefaultInstance();
  }

  @Override
  public Effect<Empty> create(
      CustomerDomain.CustomerState currentState, CustomerApi.Customer customer) {
    CustomerDomain.CustomerState customerState =
        CustomerDomain.CustomerState.newBuilder()
            .setCustomerId(customer.getCustomerId())
            .setEmail(customer.getEmail())
            .setName(customer.getName())
            .setAddress(
                customer.hasAddress()
                    ? CustomerDomain.Address.newBuilder()
                        .setStreet(customer.getAddress().getStreet())
                        .setCity(customer.getAddress().getCity())
                        .build()
                    : CustomerDomain.Address.getDefaultInstance())
            .build();
    return effects().updateState(customerState).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<CustomerApi.Customer> get(
      CustomerDomain.CustomerState currentState, CustomerApi.GetCustomer getCustomer) {
    CustomerApi.Customer customer =
        CustomerApi.Customer.newBuilder()
            .setCustomerId(currentState.getCustomerId())
            .setEmail(currentState.getEmail())
            .setName(currentState.getName())
            .setAddress(
                currentState.hasAddress()
                    ? CustomerApi.Address.newBuilder()
                        .setStreet(currentState.getAddress().getStreet())
                        .setCity(currentState.getAddress().getCity())
                        .build()
                    : CustomerApi.Address.getDefaultInstance())
            .build();
    return effects().reply(customer);
  }
}

File: samples/scala-valueentity-store/src/main/scala/store/customer/domain/CustomerValueEntity.scala
package store.customer.domain

import com.google.protobuf.empty.Empty
import kalix.scalasdk.valueentity.ValueEntity
import kalix.scalasdk.valueentity.ValueEntityContext
import store.customer.api

class CustomerValueEntity(context: ValueEntityContext) extends AbstractCustomerValueEntity {

  override def emptyState: CustomerState = CustomerState.defaultInstance

  override def create(currentState: CustomerState, customer: api.Customer): ValueEntity.Effect[Empty] = {
    val customerState = CustomerState(
      customerId = customer.customerId,
      email = customer.email,
      name = customer.name,
      address = customer.address.map(address => Address(street = address.street, city = address.city)))
    effects.updateState(customerState).thenReply(Empty.defaultInstance)
  }

  override def get(currentState: CustomerState, getCustomer: api.GetCustomer): ValueEntity.Effect[api.Customer] = {
    val customer = api.Customer(
      customerId = currentState.customerId,
      email = currentState.email,
      name = currentState.name,
      address = currentState.address.map(address => api.Address(street = address.street, city = address.city)))
    effects.reply(customer)
  }
}
